src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -707,181 +711,181 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
-                int numBytesPerLine = ((((bih.biWidth * bih.biPlanes * bih.biBitCount) + 31) >> 5) << 2);
+                        iidata.iconColors = new uint[numColors];
+                        for (int i = 0; i < numColors; i++)
+                            iidata.iconColors[i] = bihReader.ReadUInt32();
 
-                //Determine the XOR array Size
-                int xorSize = numBytesPerLine * iconHeight;
-                iidata.iconXOR = new byte[xorSize];
-                int nread = bihReader.Read(iidata.iconXOR, 0, xorSize);
-                if (nread != xorSize)
-                {
-                    string msg = string.Format("{0} data length expected {1}, read {2}", "XOR", xorSize, nread);
-                    throw new ArgumentException(msg, nameof(stream));
-                }
+                        //XOR mask is immediately after ColorTable and its size is
+                        //icon height* no. of bytes per line
 
-                //Determine the AND array size
-                numBytesPerLine = (int)((((bih.biWidth) + 31) & ~31) >> 3);
-                int andSize = numBytesPerLine * iconHeight;
-                iidata.iconAND = new byte[andSize];
-                nread = bihReader.Read(iidata.iconAND, 0, andSize);
-                if (nread != andSize)
-                {
-                    string msg = string.Format("{0} data length expected {1}, read {2}", "AND", andSize, nread);
-                    throw new ArgumentException(msg, nameof(stream));
+                        //icon height is half of BITMAPINFOHEADER.biHeight, since it contains
+                        //both XOR as well as AND mask bytes
+                        int iconHeight = bih.biHeight / 2;
+
+                        //bytes per line should should be uint aligned
+                        int numBytesPerLine = checked((((bih.biWidth * bih.biPlanes * bih.biBitCount) + 31) >> 5) << 2);
+
+                        //Determine the XOR array Size
+                        int xorSize = checked(numBytesPerLine * iconHeight);
+                        iidata.iconXOR = new byte[xorSize];
+                        int nread = bihReader.Read(iidata.iconXOR, 0, xorSize);
+                        if (nread != xorSize)
+                        {
+                            throw new ArgumentException(SR.Format(SR.IconIvalidMaskLength, "XOR", xorSize, nread), nameof(stream));
+                        }
+
+                        //Determine the AND array size
+                        numBytesPerLine = checked((((bih.biWidth) + 31) & ~31) >> 3);
+                        int andSize = checked(numBytesPerLine * iconHeight);
+                        iidata.iconAND = new byte[andSize];
+                        nread = bihReader.Read(iidata.iconAND, 0, andSize);
+                        if (nread != andSize)
+                        {
+                            throw new ArgumentException(SR.Format(SR.IconIvalidMaskLength, "AND", andSize, nread), nameof(stream));
```

Comment:
this is one of those places where there isn't a good choice for the parameter name because we're not directly called. some of the callers passed a file or a resource name. I guess stream is fine.